runtime/lua/vim/treesitter/query.lua

```diff
@@ -4,14 +4,69 @@ local memoize = vim.func._memoize
 
 local M = {}
 
+local function is_directive(name)
+  return string.sub(name, -1) == '!'
+end
+
+---A predicate is a list of (integer|string); integer represents `capture_id`, and
+---string represents (literal) arguments to predicate. A predicate is either a guard
+---or a directive. See |treesitter-predicates| and |treesitter-directives| for more details.
+---@alias Predicate (integer|string)[]
+
+---@nodoc
+---@class ProcessedGuard
+---@field [1] string handler name
+---@field [2] boolean whether it should match
+---@field [3] Predicate
+
+---@alias ProcessedDirective Predicate
+---@alias ProcessedPattern { guards: ProcessedGuard[], directives: ProcessedDirective[] }
+
+--- Splits the predicates in query patterns into guards and directives.
+---@param patterns table<integer, Predicate[]>
+---@return table<integer, ProcessedPattern>
+local function process_patterns(patterns)
+  ---@type table<integer, ProcessedPattern>
+  local processed_patterns = {}
+
+  for k, pattern in pairs(patterns) do
+    ---@type ProcessedGuard[]
+    local guards = {}
+    ---@type ProcessedDirective[]
+    local directives = {}
+
+    for _, pred in ipairs(pattern) do
+      -- Note: tree-sitter strips the leading # from predicates for us.
+      local pred_name = pred[1]
+      ---@cast pred_name string
```

Comment:
How do we know this is safe? Is the type definition correct?

Reply:
This is what the original code did. And it also expected string.

https://github.com/neovim/neovim/blob/e1c2179dd93ed2cd787b1cd016606b1901a1acfe/runtime/lua/vim/treesitter/query.lua#L772-L777